src/storage/store/list_column.cpp

```diff
@@ -343,46 +342,91 @@ void ListColumn::prepareCommitForChunk(Transaction* transaction, node_group_idx_
     if (isNewNodeGroup) {
         commitColumnChunkOutOfPlace(
             transaction, nodeGroupIdx, isNewNodeGroup, dstOffsets, chunk, startSrcOffset);
-    } else {
-        // we separate the commit into three parts: offset chunk commit, size column chunk commit,
-        // data column chunk
-        auto listChunk = ku_dynamic_cast<ColumnChunk*, ListColumnChunk*>(chunk);
-        sizeColumn->prepareCommitForChunk(
-            transaction, nodeGroupIdx, dstOffsets, listChunk->getSizeColumnChunk(), startSrcOffset);
-        auto dataColumnSize =
-            dataColumn->getMetadata(nodeGroupIdx, transaction->getType()).numValues;
-        auto dataColumnChunk = listChunk->getDataColumnChunk();
-        auto numListsToAppend = std::min(chunk->getNumValues(), (uint64_t)dstOffsets.size());
-        auto dataSize = 0u;
-        auto startListOffset = listChunk->getListStartOffset(startSrcOffset);
-        std::vector<common::offset_t> dstOffsetsInDataColumn;
-        for (auto i = 0u; i < numListsToAppend; i++) {
-            for (auto j = 0u; j < listChunk->getListSize(startSrcOffset + i); j++) {
-                dstOffsetsInDataColumn.push_back(dataColumnSize + dataSize++);
-            }
+        return;
+    }
+    // we first check if we can in place commit data column chunk
+    // if we can not in place commit data column chunk, we will out place commit offset/size/data
+    // column chunk otherwise, we commit data column chunk in place and separately commit
+    // offset/size column chunk
+    auto listChunk = ku_dynamic_cast<ColumnChunk*, ListColumnChunk*>(chunk);
+    auto dataColumnSize = dataColumn->getMetadata(nodeGroupIdx, transaction->getType()).numValues;
+    auto dataColumnChunk = listChunk->getDataColumnChunk();
+    auto numListsToAppend = std::min(chunk->getNumValues(), (uint64_t)dstOffsets.size());
+    auto dataSize = 0u;
+    auto startListOffset = listChunk->getListStartOffset(startSrcOffset);
+    std::vector<common::offset_t> dstOffsetsInDataColumn;
+    for (auto i = 0u; i < numListsToAppend; i++) {
+        for (auto j = 0u; j < listChunk->getListSize(startSrcOffset + i); j++) {
+            dstOffsetsInDataColumn.push_back(dataColumnSize + dataSize++);
         }
+    }
+    bool dataColumnCanCommitInPlace = dataColumn->canCommitInPlace(
+        transaction, nodeGroupIdx, dstOffsetsInDataColumn, dataColumnChunk, startListOffset);
+    if (!dataColumnCanCommitInPlace) {
+        commitColumnChunkOutOfPlace(
+            transaction, nodeGroupIdx, isNewNodeGroup, dstOffsets, chunk, startSrcOffset);
+    } else {
         dataColumn->prepareCommitForChunk(
```

Comment:
One optimization can be done here is to directly call out-of-place commit here for `dataColumn` as it can skip the redundant `canCommitInPlace`.